Update JsonDicomConverter.cs

nutzlastfan opened the pull request:
added  System.Globalization.CultureInfo.InvariantCulture to  double.Parse in WriteVR_Default
for fixing conversion of decimal in other culture settings (test with german)

nutzlastfan committed:
Update JsonDicomConverter.cs

added  System.Globalization.CultureInfo.InvariantCulture to  double.Parse in WriteVR_Default
for fixing conversion of decimal in other culture settings (test with german)

Zaid-Safadi commented:
Thanks for the PR @nutzlastfan!

Do you have a sample (anonymized)image that can be included in the sampleImages folder to run this [unit test](https://github.com/DICOMcloud/DICOMcloud/blob/development/UnitTests/DICOMcloud.Dicom.UnitTest/DicomDatasetConvertersTests.cs) against? 

If not, I don't mind updating one of the existing images to include this case. This way we can also validate if the XML converter needs to have a similar fix.

nutzlastfan commented:
I have, how should I send it to you ?

Zaid-Safadi commented:
You can update your PR by adding the anonymized image [here](https://github.com/DICOMcloud/DICOMcloud/tree/development/Resources/SampleImages) and run the unit test or if you have a DICOM editor, grab one of the images in that folder and update the value on an element.

nutzlastfan committed:
Add files via upload

CT image for decimal-conversion testing

nutzlastfan committed:
Update JsonDicomConverter.cs

ReadDefaultVr added System.Globalization.CultureInfo.InvariantCulture to string conversion

nutzlastfan committed:
Update JsonDicomConverter.cs

nutzlastfan committed:
Update DicomDatasetConvertersTests.cs

nutzlastfan committed:
Update DicomHelpers.cs

nutzlastfan commented:
Ok, a third ct test image is uploaded. The decimal conversion problem is reproducible with all three,
and hopefully fixed in the json converter.

The third picture also brought a new problem when "umlaute" are present. 
I added  System.Text.Encoding.Default to ReadDefaultVr to fix this.
dataset.AddOrUpdate<string> ( vr, tag, System.Text.Encoding.Default ,values.ToArray ( ) );  

 Assert.AreEqual had no error in the end. I think it would be best to confirm this again with a fully english windows in installation.

Update:
When returning a query result from Database characters like "umlaute" are not added to the dicomdataset in a way that would preserve the encoding.

I added System.Text.Encoding.Default parameter to
ReadTags:
 CurrentData.CurrentDs.AddOrUpdate<string>(dicomTag, System.Text.Encoding.Default , (string) value);

and UpdateDsPersonName:
CurrentData.CurrentDs.AddOrUpdate( personName.Key,System.Text.Encoding.Default , personName.Value.ToString());

to fix this.

nutzlastfan committed:
Update QueryResponseBuilder.cs

System.Text.Encoding.Default parameter to
ReadTags:
 CurrentData.CurrentDs.AddOrUpdate<string>(dicomTag, System.Text.Encoding.Default , (string) value);
UpdateDsPersonName:
CurrentData.CurrentDs.AddOrUpdate( personName.Key,System.Text.Encoding.Default , personName.Value.ToString());

nutzlastfan commented:
Update:
When returning a query result from Database characters like "umlaute" are not added to the dicomdataset in a way that would preserve the encoding.

I added System.Text.Encoding.Default parameter to
ReadTags:
 CurrentData.CurrentDs.AddOrUpdate<string>(dicomTag, System.Text.Encoding.Default , (string) value);

and UpdateDsPersonName:
CurrentData.CurrentDs.AddOrUpdate( personName.Key,System.Text.Encoding.Default , personName.Value.ToString());

to fix this.

Zaid-Safadi commented on UnitTests/DICOMcloud.Dicom.UnitTest/DicomDatasetConvertersTests.cs:
```diff
@@ -88,8 +88,11 @@ public void ConvertToXml ( )
 
         [TestMethod]
-        public void ConvertToJson()
+        public void ConvertToJson(Boolean SkipLargeTags = false)
```
Hi @nutzlastfan, I have a few questions about this:
How do you pass this with "true"? 
when I try to run this test in Visual Studio 2019 it doesn't let me run it anymore if you are passing a parameter. 

If there is a special test case or the new DICOM file require this, then probably create a separate unit test clarifying what you are trying to do

nutzlastfan replied:
I changed it back, i kept      DicomDataset sourceDS = DicomFile.Open ( file, FileReadOption.Default).Dataset ; in so you can manualy change the option.
I tried some other files in the test, also some dicom video files so skiplargetags
comes in handy if you only want to compare the metadata.
Next time I will make an extra test.

Zaid-Safadi commented on UnitTests/DICOMcloud.Dicom.UnitTest/Helpers/DicomHelpers.cs:
```diff
@@ -31,9 +33,14 @@ public DicomHelpers ( )
+            string baseFolder = System.AppDomain.CurrentDomain.BaseDirectory ;
+
+
+            var pathRegex = new System.Text.RegularExpressions.Regex(@"\\bin(\\x86|\\x64)?\\(Debug|Release)$", System.Text.RegularExpressions.RegexOptions.Compiled);
```
This is not working as expected and the directory returned from this function is now causing all tests to fail. 
What was the problem with the original code?

nutzlastfan replied:
I changed it back for now, the original was giving me one folder above the solution so all the paths didn't work. This was working for me.

Zaid-Safadi commented on DICOMcloud/DICOMcloud/JsonDicomConverter.cs:
```diff
@@ -483,13 +492,32 @@ int level
+
+            List<string> values = new List<string>();
+
+            while (reader.Read() && reader.TokenType == JsonToken.StartArray)
```
Can you please explain more what is the problem you are trying to solve here and why the DS VR needs a special handling?
Anyway we can test this update?

nutzlastfan replied:
Those values are affected by local culture settings, so in my case it would strip the point from a decimal completly or in the other case exchange it for a comma. So for the decimal case the value can be directly added to the dataset after this. On the other hand for the default string case in ReadDefaultVr I added  System.Text.Encoding.Default to dataset.add this should handle culture specific characters but not the conversion of decimals.

Zaid-Safadi reviewed (commented):
@nutzlastfan many thanks for your proposed changes and I believe I understand the general two problems you are trying to fix in the code but have I questions about some specifics. 

Also, the unit tests are now failing so if you can clarify the questions and ensure the unit tests are successful then I'll go ahead and merge this PR.

nutzlastfan committed:
Update DicomDatasetConvertersTests.cs

nutzlastfan committed:
Update DicomHelpers.cs

Zaid-Safadi commented:
@nutzlastfan I am troubleshooting a couple of items that are causing the unit tests to fail:

1) The attribute "Number of Frames" is being added when generating the DICOM dataset from JSON or XML. It seems that fo-dicom add this by default with value 1 if attribute doesn't exist. I don't think there is a way around this. The other 2 sample images have this attribute so I might just update your image to include the attribute and make the test work. Let me know if you have other suggestions.

![JSON-diff](https://user-images.githubusercontent.com/10813109/73715654-aef3fc00-46e2-11ea-9ec9-676fd8df93b4.PNG)

![xml-diff-2](https://user-images.githubusercontent.com/10813109/73715688-c8954380-46e2-11ea-8695-2b732dd37c64.PNG)

2) Another issue is with the XML converter not writing/reading the "umlaute" character, I'm working on a fix by specifying different XML encoding per Character-set

![xml-diff-1](https://user-images.githubusercontent.com/10813109/73715661-b4514680-46e2-11ea-8524-3503160163ce.PNG)

nutzlastfan committed:
Update XmlDicomConverter.cs

 public string Convert ( DicomDataset ds )
changed    result = Encoding.Default.GetString ( ms.ToArray ( ) ) ; to Encoding.UTF8.GetString for dealing with character issues (umlaute)

nutzlastfan commented:
1. i agree, i have the exact same result. Then to add the tag would be the solution, another working one is to use FileReadOption.SkipLargeTags option when opening the file,
then fo-dicom dosen't have the data to create the tag and only the metadata is compared.

2. I changed the default encoding from ansi to utf8 and added encoding.deafult to ds.addorupdate, tests are working, but it would be better to ckeck twice
because i don't really use the xml part at the moment so i have only the tests to go by.

nutzlastfan committed:
Update XmlDicomConverter.cs

added System.Text.Encoding.Default to string operations of ds.addorupdate

nutzlastfan committed:
Update JsonDicomConverter.cs

nutzlastfan committed:
Update XmlDicomConverter.cs

nutzlastfan committed:
Update JsonDicomConverter.cs

Zaid-Safadi commented:
@nutzlastfan I am going to merge this into its own branch so we can better collaborate on it. Please create two issues to describe the problems you addressed in this PR and I'll cherry pick and merge into development to close them out while working on the unit tests and XML converter
